src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnDecoder.Integer.cs

```diff
@@ -0,0 +1,748 @@
+        ///   <see cref="TagClass.Universal"/>, but
+        ///   <paramref name="expectedTag"/>.<see cref="Asn1Tag.TagValue"/> is not correct for
+        ///   the method.
+        /// </exception>
+        [CLSCompliant(false)]
+        public static bool TryReadUInt64(
+            ReadOnlySpan<byte> source,
+            AsnEncodingRules ruleSet,
+            out ulong value,
+            out int bytesConsumed,
+            Asn1Tag? expectedTag = null)
+        {
+            return TryReadUnsignedInteger(
+                source,
+                ruleSet,
+                sizeof(ulong),
+                expectedTag ?? Asn1Tag.Integer,
+                UniversalTagNumber.Integer,
+                out value,
+                out bytesConsumed);
+        }
+
+        private static ReadOnlySpan<byte> GetIntegerContents(
+            ReadOnlySpan<byte> source,
+            AsnEncodingRules ruleSet,
+            Asn1Tag expectedTag,
+            UniversalTagNumber tagNumber,
+            out int bytesConsumed)
+        {
+            // T-REC-X.690-201508 sec 8.3.1
+            ReadOnlySpan<byte> contents = GetPrimitiveContentSpan(
+                source,
+                ruleSet,
+                expectedTag,
+                tagNumber,
+                out int consumed);
+
+            // T-REC-X.690-201508 sec 8.3.1
+            if (contents.IsEmpty)
+            {
+                throw new AsnContentException();
+            }
+
+            // T-REC-X.690-201508 sec 8.3.2
+            if (contents.Length > 1)
+            {
+                ushort bigEndianValue = BinaryPrimitives.ReadUInt16BigEndian(contents);
+                const ushort RedundancyMask = 0b1111_1111_1000_0000;
+                ushort masked = (ushort)(bigEndianValue & RedundancyMask);
+
+                // If the first 9 bits are all 0 or are all 1, the value is invalid.
+                if (masked == 0 || masked == RedundancyMask)
+                {
+                    throw new AsnContentException();
+                }
+            }
+
+            bytesConsumed = consumed;
+            return contents;
+        }
+
+        private static bool TryReadSignedInteger(
+            ReadOnlySpan<byte> source,
+            AsnEncodingRules ruleSet,
+            int sizeLimit,
+            Asn1Tag expectedTag,
+            UniversalTagNumber tagNumber,
+            out long value,
+            out int bytesConsumed)
+        {
+            Debug.Assert(sizeLimit <= sizeof(long));
+
+            ReadOnlySpan<byte> contents = GetIntegerContents(
+                source,
+                ruleSet,
+                expectedTag,
+                tagNumber,
+                out int consumed);
+
+            if (contents.Length > sizeLimit)
+            {
+                value = 0;
+                bytesConsumed = 0;
+                return false;
+            }
+
+            bool isNegative = (contents[0] & 0x80) != 0;
+            long accum = isNegative ? -1 : 0;
```

Comment:
Nit: `long accum = (sbyte)contents[0];`, which will sign-extend the 8-bit `contents[0]` value to 64 bits. Then we can drop line 449 entirely, and on line 452 start the _for_ loop at _i = 1_.